p2p/message.go

```diff
@@ -102,73 +103,257 @@ func (sm *StatusMessage) Encode() ([]byte, error) {
 	if err != nil {
 		return enc, err
 	}
-	return append([]byte{StatusMsg}, enc...), nil
+	return append([]byte{StatusMsgType}, enc...), nil
 }
 
 // Decodes the message into a StatusMessage, it assumes the type byte has been removed
-func (sm *StatusMessage) Decode(msg []byte) error {
-	_, err := scale.Decode(msg, sm)
+func (sm *StatusMessage) Decode(r io.Reader) error {
+	sd := scale.Decoder{Reader: r}
+	_, err := sd.Decode(sm)
 	return err
 }
 
+// for optionals, if first byte is 0, then it is None
+// otherwise it is Some
 type BlockRequestMessage struct {
-	Id            uint32
+	Id            uint64
 	RequestedData byte
-	StartingBlock []byte      // first byte 0 = block hash (32 byte), first byte 1 = block number (int64)
-	EndBlockHash  common.Hash // optional
+	StartingBlock []byte              // first byte 0 = block hash (32 byte), first byte 1 = block number (int64)
+	EndBlockHash  common.OptionalHash // optional [32]byte
 	Direction     byte
-	Max           uint32 // optional
+	Max           common.OptionalUint32 // optional uint32
 }
 
 // String formats a BlockRequestMessage as a string
 func (bm *BlockRequestMessage) String() string {
-	return fmt.Sprintf("BlockRequestMessage Id=%d RequestedData=%d StartingBlock=0x%x EndBlockHash=0x%x Direction=%d Max=%d",
-		bm.Id,
-		bm.RequestedData,
-		bm.StartingBlock,
-		bm.EndBlockHash,
-		bm.Direction,
-		bm.Max)
+	if bm.EndBlockHash.Exists && bm.Max.Exists {
```

Comment:
You can save yourself from this by implementing `String()` methods for the optional types. They will always be included but they can be shown as nil or w/e

Reply:
done